server/src/main/java/org/apache/cloudstack/diagnosis/RemoteDiagnosticsServiceImpl.java

```diff
@@ -0,0 +1,121 @@
+
+
+    @Override
+    public RemoteDiagnosticsResponse executeDiagnosisToolInSystemVm(final RemoteDiagnosticsCmd cmd) throws AgentUnavailableException,
+            InvalidParameterValueException {
+        final Long systemVmId = cmd.getId();
+        final VMInstanceVO systemVm = vmInstanceDao.findByIdTypes(systemVmId, VirtualMachine.Type.ConsoleProxy,
+                VirtualMachine.Type.DomainRouter, VirtualMachine.Type.SecondaryStorageVm);
+        if (systemVm == null){
+            s_logger.error("Unable to find a system virtual machine with id " + systemVmId);
+            throw new InvalidParameterValueException("Unable to find a system virtual machine with id " + systemVmId);
+        }
+
+        final Long hostId = systemVm.getHostId();
+        if (hostId == null){
+            s_logger.warn("Unable to find host for virtual machine instance " + systemVm.getInstanceName());
+            throw new InvalidParameterValueException("Unable to find host for virtual machine instance " + systemVm.getInstanceName());
+        }
+
+        final String diagnosisCommandType = cmd.getType();
+        final String destinationIpAddress = cmd.getAddress();
+        final String optionalArgunments = cmd.getOptionalArguments();
+
+        String remoteCommand = setupRemoteCommand(diagnosisCommandType, destinationIpAddress, optionalArgunments);
+
+        final ExecuteDiagnosisCommand command = new ExecuteDiagnosisCommand(remoteCommand);
+        command.setAccessDetail(NetworkElementCommand.ROUTER_IP, routerControlHelper.getRouterControlIp(systemVm.getId()));
+        command.setAccessDetail(NetworkElementCommand.ROUTER_NAME, systemVm.getInstanceName());
+
+        final String commandPassed = command.getSrciptArguments();
+
+        Answer origAnswer;
+        try{
+            origAnswer = agentManager.send(hostId, command);
+        } catch (final OperationTimedoutException e) {
+            s_logger.warn("Timed Out", e);
+            throw new AgentUnavailableException("Unable to send commands to virtual machine ", hostId, e);
+        }
+
+        ExecuteDiagnosisAnswer answer = null;
+        if (origAnswer instanceof ExecuteDiagnosisAnswer) {
+            answer = (ExecuteDiagnosisAnswer) origAnswer;
+        }
+
+        return createRemoteDiagnosisResponse(answer,commandPassed);
+    }
+
+    private static RemoteDiagnosticsResponse createRemoteDiagnosisResponse(ExecuteDiagnosisAnswer answer, String commandPaased){
+        RemoteDiagnosticsResponse response = new RemoteDiagnosticsResponse();
+        response.setResult(answer.getResult());
+        response.setDetails(answer.getDetails());
+        response.setNetworkCommand(commandPaased);
+        return response;
+    }
+
+    private static String setupRemoteCommand(String diagnosisType, String destinationIpAddress, String optionalArguments){
+        if (optionalArguments != null){
+            return String.format("%s %s", diagnosisType, destinationIpAddress+" "+optionalArguments);
```

Comment:
so the optional arguments are always last. for some commands that will exclude options. thought the trend is to allow for any argument on any place in teh command line. This is possible place for augmentation.

Reply:
I am working on refactoring this entire method during migration to python based script. I aim at parsing all arguments passed for each command type.